File: python/lombokalgoritma/graph/topological_sort.py
# LombokAlgoritma — topological sort, Kahn (SPEC §9.7)
# SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
from __future__ import annotations

from .types import Graph, build_adj_list


def topological_sort(g: Graph) -> list[int]:
    """Kahn with a FIFO queue seeded by zero in-degree nodes ascending; ``[]`` on a cycle."""
    adj = build_adj_list(g)
    n = len(adj)
    indegree = [0] * n
    for _, v, _ in g.edges:
        indegree[v] += 1
    queue = [i for i in range(n) if indegree[i] == 0]
    head = 0
    while head < len(queue):
        u = queue[head]
        head += 1
        for v, _ in adj[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                queue.append(v)
    return queue if len(queue) == n else []

File: typescript/src/graph/topological-sort.ts
// LombokAlgoritma — Topological sort (Kahn)
// SPDX-License-Identifier: Apache-2.0 OR MIT — @codinglombok
import { type Graph, buildAdjList } from './types.js';

/**
 * Kahn's algorithm with a FIFO queue seeded with the zero-in-degree nodes in ascending order;
 * neighbours are released in edge-list order. Returns `[]` when the graph has a cycle.
 * O(V + E). SPEC §6.8.
 */
export function topologicalSort(g: Graph): number[] {
  const adj = buildAdjList(g);
  const indegree = new Array<number>(g.nodes).fill(0);
  for (const e of g.edges) indegree[e.to] = (indegree[e.to] as number) + 1;
  const queue: number[] = [];
  for (let i = 0; i < g.nodes; i++) if (indegree[i] === 0) queue.push(i);
  for (let head = 0; head < queue.length; head++) {
    const u = queue[head] as number;
    for (const { to } of adj[u] ?? []) {
      indegree[to] = (indegree[to] as number) - 1;
      if (indegree[to] === 0) queue.push(to);
    }
  }
  return queue.length === g.nodes ? queue : [];
}
